modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-service/src/main/java/com/liferay/commerce/shop/by/diagram/service/impl/CPDefinitionDiagramEntryServiceImpl.java

```diff
@@ -43,10 +43,101 @@
 public class CPDefinitionDiagramEntryServiceImpl
 	extends CPDefinitionDiagramEntryServiceBaseImpl {
 
-	/**
-	 * NOTE FOR DEVELOPERS:
-	 *
-	 * Never reference this class directly. Always use <code>com.liferay.commerce.shop.by.diagram.service.CPDefinitionDiagramEntryServiceUtil</code> to access the cp definition diagram entry remote service.
-	 */
+	@Override
+	public CPDefinitionDiagramEntry addCPDefinitionDiagramEntry(
+			long userId, String cpInstanceUuid, long cProductId,
+			long cpDefinitionId, int number, String sku, int quantity,
+			boolean diagram, ServiceContext serviceContext)
+		throws PortalException {
+
+		_cpDefinitionModelResourcePermission.check(
+			getPermissionChecker(), cpDefinitionId, ActionKeys.UPDATE);
+
+		return cpDefinitionDiagramEntryLocalService.addCPDefinitionDiagramEntry(
+			userId, cpInstanceUuid, cProductId, cpDefinitionId, number, sku,
+			quantity, diagram, serviceContext);
+	}
+
+	@Override
+	public void deleteCPDefinitionDiagramEntry(long cpDefinitionDiagramEntryId)
+		throws PortalException {
+
+		CPDefinitionDiagramEntry cpDefinitionDiagramEntry =
+			cpDefinitionDiagramEntryLocalService.getCPDefinitionDiagramEntry(
+				cpDefinitionDiagramEntryId);
+
+		_cpDefinitionModelResourcePermission.check(
+			getPermissionChecker(),
+			cpDefinitionDiagramEntry.getCPDefinitionId(), ActionKeys.UPDATE);
```

Comment:
```suggestion
	public void deleteCPDefinitionDiagramEntry(long cpDefinitionDiagramEntryId)
		throws PortalException {

		CPDefinitionDiagramEntry cpDefinitionDiagramEntry =
			cpDefinitionDiagramEntryLocalService.getCPDefinitionDiagramEntry(
				cpDefinitionDiagramEntryId);

		_cpDefinitionModelResourcePermission.check(
			getPermissionChecker(),
			cpDefinitionDiagramEntry.getCPDefinitionId(), ActionKeys.DELETE);
```

Reply:
As I said to Igor, these actions should be possible only if a product can be updated, since it's all part of the edit screen of the product.